Add script for populating test DB with experiences

hlibbabii opened the pull request:
fixes  [#1699](https://github.com/Trustroots/trustroots/issues/1669)

#### Testing instructions

- run `npm seed`
- check the DB is populated with references

simison commented:
> ReferenceThread

No worries, this can be ignored for now. It's a different feature; there used to be ["are messages in spirit of Trustroots?"](https://github.com/Trustroots/trustroots/blob/a6114484a33a0275252261c81fba7c06b7aacdd1/modules/references-thread/client/views/directives/tr-reference-thread.client.view.html#L8-L28) question next to message threads but in Angular->React refactor it slipped away (we could add it back, should be simple).

simison commented on bin/fillTestData/util.js:
```diff
@@ -0,0 +1,22 @@
+ * @returns random integer between 0 and max - 1
+ */
+exports.random = function random(max) {
+  return Math.floor(Math.random() * max);
```
lodash.random might be enough here: https://lodash.com/docs/4.17.15#random

hlibbabii commented on package.json:
```diff
@@ -39,10 +39,11 @@
     "reformat-files": "prettier --ignore-path .eslintignore --write",
     "preseed": "npm run check-environment",
     "prestart": "npm run check-environment",
+    "seed:experiences": "node ./bin/fillTestData/Experiences.js",
```
Currently, when running `npm seed:experiences` you get an error because the script requires parameters. The same for messages, circles, and users. If we want to be able to run each `npm seed:<collection-name>` separately, we could fix all these scripts together by defining default parameters with yargs or something like this.

simison replied:
> we could fix all these scripts together by defining default parameters with yargs or something like this.

Can be a concern for follow-up PR.

hlibbabii commented on bin/fillTestData/Experiences.js:
```diff
@@ -0,0 +1,342 @@
+  },
+).argv;
+
+const profileType = {
```
I think I have over-engineered things here a little, and also to have these magic numbers is not nice and robust enough. But in the end, I thought for a 'dev' task it might be good enough for now. Let me know what you think.

simison replied:
You could always add a comment above the object to elaborate more on numbers if you feel like it. :-)

hlibbabii commented on bin/fillTestData/Experiences.js:
```diff
@@ -0,0 +1,342 @@
+  }
+
+  function populateRandomly(expMatrix, averageExpNumber, replyRate) {
+    function getExperienceSharingRate(nUsers, averageExpNumber, replyRate) {
```
Also would be glad to receive any critique regarding the way I structure the JS code, for example, regarding function nesting like this.

simison replied:
It does affect readability in my opinion but it's not a biggie in a script like this. I'd be inclined to leave as-is since it works. :-)

simison commented on bin/fillTestData/Messages.js:
```diff
@@ -171,7 +149,7 @@ function seedThreads() {
                 const message = new Message();
 
-                message.created = addDays(Date.now(), -depth + 1);
+                message.created = util.addDays(Date.now(), -depth + 1);
```
Just FYI, we also have `moment` available which has some convenience methods. Moment.js isn't great at frontend (because of bloat) but kinda fine at scripts like this.

Fine using the simple utility as well though. :-) It might just behave oddly on the last day of the month?

simison replied:
Edit: oh I see you just moved the existing utility to another file. Never mind!

simison reviewed (approved):
I tested `npm run seed:experiences 10 0.8 -- --force` and got:

```
Running with --force option. Removing existing Experiences ...
Profile with many experiences (username): 100brettaufderhar
Saving 15374 experiences ...

<--- Last few GCs --->

[11097:0x102d57000]    69117 ms: Scavenge 2046.0 (2050.2) -> 2045.5 (2051.0) MB, 492.0 / 0.0 ms  (average mu = 0.175, current mu = 0.001) allocation failure
[11097:0x102d57000]    70874 ms: Mark-sweep 2046.5 (2051.0) -> 2045.7 (2050.7) MB, 1755.8 / 0.0 ms  (average mu = 0.201, current mu = 0.220) allocation failure scavenge might not succeed


<--- JS stacktrace --->

==== JS stack trace =========================================

    0: ExitFrame [pc: 0x1009dcbd9]
Security context: 0x3443f6ec08d1 <JSObject>
    1: /* anonymous */ [0x34430233ffb9] [/Users/mikael/Documents/trustroots/bin/fillTestData/Experiences.js:~306] [pc=0x18c49596db14](this=0x344375bc03f9 <JSGlobal Object>,0x34431353ffd1 <JSArray[4054]>,2796)
    2: arguments adaptor frame: 3->2
    3: map [0x3443f6ed8ec9](this=0x344337e3ffa9 <JSArray[4054]>,0x34430233ffb9 <JSFunction (sfi = 0x344389afe2a9)>)
 ...

FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
 1: 0x1011d1cc5 node::Abort() (.cold.1) [/usr/local/bin/node]
 2: 0x10009f919 node::Abort() [/usr/local/bin/node]
 3: 0x10009fa7f node::OnFatalError(char const*, char const*) [/usr/local/bin/node]
 4: 0x1001e38b7 v8::Utils::ReportOOMFailure(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 5: 0x1001e3857 v8::internal::V8::FatalProcessOutOfMemory(v8::internal::Isolate*, char const*, bool) [/usr/local/bin/node]
 6: 0x10036b9e5 v8::internal::Heap::FatalProcessOutOfMemory(char const*) [/usr/local/bin/node]
 7: 0x10036d25a v8::internal::Heap::RecomputeLimits(v8::internal::GarbageCollector) [/usr/local/bin/node]
 8: 0x100369c8c v8::internal::Heap::PerformGarbageCollection(v8::internal::GarbageCollector, v8::GCCallbackFlags) [/usr/local/bin/node]
 9: 0x100367a8e v8::internal::Heap::CollectGarbage(v8::internal::AllocationSpace, v8::internal::GarbageCollectionReason, v8::GCCallbackFlags) [/usr/local/bin/node]
10: 0x10037395a v8::internal::Heap::AllocateRawWithLightRetry(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
11: 0x1003739e1 v8::internal::Heap::AllocateRawWithRetryOrFail(int, v8::internal::AllocationType, v8::internal::AllocationOrigin, v8::internal::AllocationAlignment) [/usr/local/bin/node]
12: 0x1003413aa v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationType, v8::internal::AllocationOrigin) [/usr/local/bin/node]
13: 0x1006937b8 v8::internal::Runtime_AllocateInYoungGeneration(int, unsigned long*, v8::internal::Isolate*) [/usr/local/bin/node]
14: 0x1009dcbd9 Builtins_CEntry_Return1_DontSaveFPRegs_ArgvOnStack_NoBuiltinExit [/usr/local/bin/node]
15: 0x18c49596db14
16: 0x10095bbdc Builtins_ArgumentsAdaptorTrampoline [/usr/local/bin/node]
zsh: abort      npm run seed:experiences 10 0.8 -- --force
```

What do you think about these changes:
- Either default on overwriting experienced (i.e. no need for `force` CLI arg), or...
- Change `--force` to `--re-create-references`, `--force-references` or similar. `force` is already an NPM argument so that can be confusing, and I get a warning from NPM:
  > npm WARN using --force I sure hope you know what you are doing.
- You need a double dash to pass the argument to CLI script in NPM: `npm run example -- --force`: otherwise it passes it to `npm`, and not to the script.  

Feel free to merge though, I think all the above things can be follow-up changes as well. One option could be to remove the seed script from the main `seed` command for now until further changes. Up to you!

Thanks for working on this!

hlibbabii commented:
@simison so I changed the flag to `--re-create`, and also fixed the out-of-memory issue by creating a pool of promises instead of trying to save thousands of experiences into the DB at the same time

simison commented on bin/fillTestData/Experiences.js:
```diff
@@ -0,0 +1,343 @@
+const yargs = require('yargs');
+const faker = require('faker');
+const mongoose = require('mongoose');
+const PromisePool = require('es6-promise-pool');
```
We also have [`async`](https://www.npmjs.com/package/async) available in the codebase, which has methods like [`queue`](https://caolan.github.io/async/v3/docs.html#queue).

Not opposed using the pool helper but just pointing out existing things in case you'd wanna consider those. :-)

Simpler approach (that might be enough here?) is [Mongo's `insertMany()`](https://mongoosejs.com/docs/api.html#model_Model.insertMany) that has some built-in controls for parallelization.

hlibbabii replied:
Cool, thanks for giving these hints! Using async/queue now.

simison commented on bin/fillTestData/Experiences.js:
```diff
@@ -0,0 +1,343 @@
+
+      function saveExperience(experience) {
+        return new Promise((resolve, reject) => {
+          experience.save(err => {
```
Would returning `experience.save()` here directly work without callbacks since I think it's a Promise already?

simison commented on bin/fillTestData/util.js:
```diff
@@ -0,0 +1,12 @@
+/**
+ * Adds number of days to the date and returns a new date
```
Another way would be to use `moment`:
```js
moment(date).add({ days }).toDate()
```

simison reviewed (approved):
Re-tested by running `npm run seed` —

worked great:

<img width="643" alt="image" src="https://user-images.githubusercontent.com/87168/99413839-dda03380-28fe-11eb-8d81-24abb268457a.png">

Don't consider any of my feedback as blocking merging. It's a dev task that doesn't need to be super-optimized. ;-)

simison commented:
Hm, it did work before the queue and now I get this again:

> FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory

Fine to revert to previous version if you'd just prefer to get it merged. :-)

hlibbabii commented:
> Hm, it did work before the queue and now I get this again:
> 
> > FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory
> 
> Fine to revert to previous version if you'd just prefer to get it merged. :-)

Hmm, so strange, it used to work also with the latest version, but now when I rerun it, it doesn't work for me either:(

simison commented:
Oh, never mind, it's actually fataling on message thread generation (`npm run seed:messages 2000 10`).  😓 

I'll try clearing out Messages collection (57094 records) to see if it helps.

hlibbabii commented:
> Oh, never mind, it's actually fataling on message thread generation (`npm run seed:messages 2000 10`). 😓

Also on users `npm run seed:users 1000`

simison reviewed (approved):
Aight! Worked great on empty database. I think that's all what we need. :-)

```
node bin/db-drop.js
npm run seed
```
